avm/ptn/dev-ops/cicd-agents-and-runners/main.bicep

```diff
@@ -0,0 +1,1030 @@
+                    secureValue: any(selfHostedConfig.personalAccessToken)
+                  }
+                ]
+              : [
+                  {
+                    name: 'AZP_URL'
+                    value: devOpsOrgURL
+                  }
+                  {
+                    name: 'AZP_POOL'
+                    value: selfHostedConfig.agentsPoolName
+                  }
+                  {
+                    name: 'AZP_AGENT_NAME'
+                    value: any(selfHostedConfig.?agentName ?? 'agent-${i}')
+                  }
+                  {
+                    name: 'AZP_TOKEN'
+                    secureValue: any(selfHostedConfig.personalAccessToken)
+                  }
+                ]
+          }
+        }
+      ]
+      ipAddressPorts: [
+        {
+          port: 80
+          protocol: 'TCP'
+        }
+      ]
+    }
+  }
+]
+
+module acaJob 'br/public:avm/res/app/job:0.4.0' = if (contains(computeTypes, 'azure-container-app')) {
+  name: '${namingPrefix}acaJob'
+  dependsOn: [
+    taskRun
+    runPlaceHolderAgent
+  ]
+  params: {
+    name: '${namingPrefix}-${uniqueString(resourceGroup().id)}-acajob'
+    location: location
+    managedIdentities: {
+      userAssignedResourceIds: [
+        userAssignedIdentity.outputs.resourceId
+      ]
+    }
+    roleAssignments: [
+      {
+        principalId: userAssignedIdentity.outputs.principalId
+        roleDefinitionIdOrName: 'Contributor'
+        principalType: 'ServicePrincipal'
+        name: guid(userAssignedIdentity.outputs.principalId, 'Contributor', resourceGroup().id)
+      }
+    ]
+    triggerType: 'Event'
+    replicaRetryLimit: 3
+    replicaTimeout: 1800
+    eventTriggerConfig: {
+      parallelism: 1
+      replicaCompletionCount: 1
+      scale: {
+        maxExecutions: 100
+        minExecutions: 0
+        pollingInterval: 30
+        rules: selfHostedConfig.selfHostedType == 'github' ? acaGitHubRules : acaAzureDevOpsRules
+      }
+    }
+    registries: [
+      {
+        server: acr.outputs.loginServer
+        identity: userAssignedIdentity.outputs.resourceId
+      }
+    ]
+    secrets: selfHostedConfig.selfHostedType == 'github' ? acaGitHubSecrets : acaAzureDevOpsSecrets
+    containers: [
+      {
+        image: '${acr.outputs.loginServer}/${selfHostedConfig.selfHostedType}-azure-container-app:latest'
+        name: 'private-runner-${selfHostedConfig.selfHostedType}-aca'
+        resources: selfHostedConfig.?azureContainerAppTarget.?resources ?? {
+          cpu: '1'
+          memory: '2Gi'
+        }
+        env: selfHostedConfig.selfHostedType == 'github' ? acaGitHubEnvVariables : acaAzureDevOpsEnvVariables
+      }
+    ]
+    environmentResourceId: appEnvironment.outputs.resourceId
+    workloadProfileName: 'consumption'
+  }
+}
+
+module acaPlaceholderJob 'br/public:avm/res/app/job:0.4.0' = if (contains(computeTypes, 'azure-container-app') && selfHostedConfig.selfHostedType == 'azuredevops') {
```

Comment:
Why the 'placeholder' naming?

Reply:
the agent itself is called a placeholder agent; it needs to exist before ay other agent can take a job.

Reply:
**This** is madness. But will resolve 😄